src/lib/pgen.rs

```diff
@@ -1,11 +1,52 @@
-use grammar::{Grammar, Symbol};
-use std::collections::{HashMap, HashSet};
-use std::hash::{Hash, Hasher};
+extern crate bit_vec;
+use self::bit_vec::BitVec;
 
-/// Generates and returns the first set for the given grammar.
+use grammar::{Grammar, NIdx, Symbol, TIdx};
+//use std::collections::{HashMap, HashSet};
+//use std::hash::{Hash, Hasher};
+
+/// Firsts stores all the first sets for a given grammar.
+#[derive(Debug)]
+pub struct Firsts {
+    // The representation is a contiguous bitfield, of (terms_len * 1) * nonterms_len. Put another
+    // way, each nonterminal has (terms_len + 1) bits, where the bit at position terms_len
+    // represents epsilon.
+    bf: BitVec,
+    terms_len: NIdx
+}
+
+impl Firsts {
+    fn new(nonterms_len: NIdx, terms_len: TIdx) -> Firsts {
+        Firsts {
+            bf        : BitVec::from_elem((nonterms_len * (terms_len + 1)) as usize, false),
+            terms_len : terms_len
+        }
+    }
+
+    /// Returns true if the firsts bit for terminal `tidx` nonterminal `nidx` is set, or
+    /// false otherwise. Bit `terms_len` represents epsilon.
+    pub fn get(&self, nidx: NIdx, tidx: TIdx) -> bool {
```

Comment:
I don't think I understand how this works. Let's say I want to know the all terminals that are contained in the first-set of `S`. What do I do?